tools/rpkg/tests/testthat/test_relational.R

```diff
@@ -275,3 +275,234 @@ test_that("Symmetric difference returns the symmetric difference", {
+    sum2 <- list(expr_reference("a"))
+    order_over_window <- rapi_rel_order(window_function, sum2)
+    expected_result <- data.frame(a=c(1:8),b=c(1, 1, 2, 2, 3, 3, 4, 4), prev_a=c(NA, 1, 2, 3, 4, 5, 6, 7))
+    res <- rel_to_altrep(order_over_window)
+    expect_equal(res, expected_result)
+})
+
+test_that("Window lead function works as expected", {
+#   select a, b, lag(a, 1) OVER () order by a
+    rel_a <- rel_from_df(con, data.frame(a=c(1:8),b=c(1, 1, 2, 2, 3, 3, 4, 4)))
+    lead <- list(expr_reference("a"))
+    window_function <- rel_window(rel=rel_a, window_function="lead", window_alias="next_a", children=lead, offset=list(expr_constant(1)))
+    sum2 <- list(expr_reference("a"))
+    order_over_window <- rapi_rel_order(window_function, sum2)
+    expected_result <- data.frame(a=c(1:8),b=c(1, 1, 2, 2, 3, 3, 4, 4), next_a=c(2, 3, 4, 5, 6, 7, 8, NA))
+    res <- rel_to_altrep(order_over_window)
+    expect_equal(res, expected_result)
+})
+
+test_that("Window function with string aggregate works", {
+#   select j, s, string_agg(s, '|') over (partition by b) from a order by j, s;
+    rel_a <- rel_from_df(con, data.frame(r=c(1, 2, 3, 4), a=c("hello", "Big", "world", "42"),b=c(1, 1, 2, 2)))
+    str_agg <- list(expr_reference("a"))
+    partitions <- list(expr_reference("b"))
+    window_function <- rel_window(rel=rel_a, window_function="string_agg", window_alias="str_concat", children=str_agg, partitions=partitions)
+    sum2 <- list(expr_reference("r", window_function))
+    order_over_window <- rapi_rel_order(window_function, sum2)
+    expected_result <- data.frame(r=c(1:4),a=c("hello", "Big", "world", "42"), b=c(1, 1, 2, 2), str_concat=c("hello,Big", "hello,Big", "world,42", "world,42"))
+    res <- rel_to_altrep(order_over_window)
+    expect_equal(res, expected_result)
+})
+
+test_that("You can project from a window function", {
+#     select j, i, sum(i) over (partition by j) from a order by 1,2
+    rel_a <- rel_from_df(con, data.frame(a=c(1:8),b=c(1, 1, 2, 2, 3, 3, 4, 4)))
+    sum <- list(expr_reference("a"))
+    partitions <- list(expr_reference("b"))
+    window_function <- rel_window(rel=rel_a, window_alias="window_result", window_function="sum", children=sum, partitions = partitions)
+    sum2 <- list(expr_reference("a"))
+    order_over_window <- rapi_rel_order(window_function, sum2)
+    proj <- rapi_rel_project(order_over_window, list(expr_reference("window_result")))
+    expected_result <- data.frame(window_result=c(3, 3, 7, 7, 11, 11, 15, 15))
+    res = rel_to_altrep(proj)
+    expect_equal(res, expected_result)
+})
+
+test_that("You can perform window functions on row_number", {
+	# select a, b, row_number() OVER () from tmp order by a;
+    rel_a <- rel_from_df(con, data.frame(a=c(8:1),b=c(1, 1, 2, 2, 3, 3, 4, 4)))
+    sum <- list(expr_reference("a"))
+    partitions <- list(expr_reference("b"))
+    window_function <- rel_window(rel=rel_a, window_function="row_number", window_alias="row_number")
```

Comment:
I would think the actual window function should be an expression no? Then you would not have the function and the children separately either?